spec/workflow/intrinsic_function/parser_spec.rb

```diff
@@ -0,0 +1,214 @@
+      expect(subject).to parse("0.0e1")
+      expect(subject).to parse("0.0e123")
+      expect(subject).to parse("0.01e1")
+      expect(subject).to parse("0.01e123")
+      expect(subject).to parse("123.123e1")
+      expect(subject).to parse("123.123e123")
+      expect(subject).to parse("123.123e-1")
+      expect(subject).to parse("123.123e-123")
+
+      expect(subject).to parse("-0.0e1")
+      expect(subject).to parse("-0.0e123")
+      expect(subject).to parse("-0.01e1")
+      expect(subject).to parse("-0.01e123")
+      expect(subject).to parse("-123.123e1")
+      expect(subject).to parse("-123.123e123")
+      expect(subject).to parse("-123.123e-1")
+      expect(subject).to parse("-123.123e-123")
+
+      expect(subject).to_not parse(".")
+      expect(subject).to_not parse(".1")
+      expect(subject).to_not parse("1.")
+      expect(subject).to_not parse("-1.")
+      expect(subject).to_not parse("1..2")
+      expect(subject).to_not parse("1.e1")
+      expect(subject).to_not parse("-1.e1")
+      expect(subject).to_not parse("01")
+      expect(subject).to_not parse("1e")
+      expect(subject).to_not parse("e")
+      expect(subject).to_not parse("e123")
+    end
+  end
+
+  describe "string" do
+    subject { described_class.new.string }
+
+    it do
+      expect(subject).to parse(%q|''|)
+      expect(subject).to parse(%q|'str'|)
+      expect(subject).to parse(%q|'str\'str'|)
+      expect(subject).to parse(%q|'str\\\\str'|)
+
+      expect(subject).to_not parse("str")
+      expect(subject).to_not parse("'str")
+      expect(subject).to_not parse("str'")
+      expect(subject).to_not parse("'str'str'")
+
+      # TODO: Many more test cases
+    end
+  end
+
+  describe "json_path" do
+    subject { described_class.new.jsonpath }
+
+    it do
+      expect(subject).to parse("$.input")
+      expect(subject).to parse("$.input.foo")
+      expect(subject).to parse("$$.input")
+      expect(subject).to parse("$$.input.foo")
+
+      expect(subject).to parse(%q|$.store.book|)
+      expect(subject).to parse(%q|$.store\.book|)
+      expect(subject).to parse(%q|$.\stor\e.boo\k|)
+      expect(subject).to parse(%q|$.store.book.title|)
+      expect(subject).to parse(%q|$.foo.\.bar|)
+      expect(subject).to parse(%q|$.foo\@bar.baz\[\[.\?pretty|)
+      expect(subject).to parse(%q|$.&Ж中.\uD800\uDF46|)
+      expect(subject).to parse(%q|$.ledgers.branch[0].pending.count|)
+      expect(subject).to parse(%q|$.ledgers.branch[0]|)
+      expect(subject).to parse(%q|$.ledgers[0][22][315].foo|)
+      expect(subject).to parse(%q|$['store']['book']|)
+      expect(subject).to parse(%q|$['store'][0]['book']|)
+
+      # TODO: These unlikely cases should work, but we need a more thorough parser.
+      expect(subject).to_not parse(%q|$['s,tore']|)
+      expect(subject).to_not parse(%q|$['s)tore']|)
+
+      expect(subject).to_not parse(%q|['input']|)
+      expect(subject).to_not parse(".input")
+      expect(subject).to_not parse("$")
```

Comment:
Hm this one is interesting, isn't `$` valid jsonpath for "return the entire input"

Reply:
Yes.  That is true and so is a bug.  I was thinking it was bad input, and forgot about that case.

Reply:
Pointing back to https://github.com/ManageIQ/floe/pull/194#discussion_r1662606992, I'm going to add using Floe::Workflow::Path in this PR, because it will handle those cases and I can add support for context at the same time automatically.

Reply:
:+1:

Reply:
Ok done.  That came out nicer than I expected - I didn't realize Floe::Workflow::Path was so nicely self-contained.